src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.CSharp.props

```diff
@@ -12,7 +12,6 @@ Copyright (c) .NET Foundation. All rights reserved.
 <Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
 
   <PropertyGroup>
-    <WarningLevel Condition=" '$(WarningLevel)' == '' ">4</WarningLevel>
```

Comment:
This is potentially controversial, but I think necessary because of the way props/projects/targets properties are ordered during import. Due to this ordering, `WarningLevel` is virtually guaranteed to be `''` at this point, so we'll set it here, which means if a user explicitly sets a WarningLevel in their project file we won't have a way of distinguishing that. This layering is part of the reason why we've moved more towards settings properties in `targets` files over the years.

I've checked and nothing seems to modify WarningLevel in this way aside from the Analyzers targets, so this seems pretty safe. Compilers and the ILLink targets consume it, but that's well after the Analyzers targets run.

Nonetheless I've reached out to some Roslyn folks to confirm.

Reply:
@baronfel Can we reach a point where WarningLevel is never set? For .NET Framework for example where we might not do anything for it in Microsoft.NET.Sdk.Analyzers.targets?

Is it safer to move the deleted line into a PropertyGroup after:

https://github.com/dotnet/sdk/blob/4a495a322321e95c0c8da15cff6c857aacf9d00a/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets#L1158

Reply:
Another way to do this would be to put the default set in `CSharp.targets` instead of `CSharp.props`, and in `CSharp.targets`

* set the warning level if not set at this point (post-project-file-import occurring),
* also set a `_WarningLevelWasSet` boolean property that we could use instead of `WarningLevel == ''` in the Analyzers target

I believe I've seen this pattern elsewhere in the SDK.

This illustrates that there should probably be a .NET Framework test as well for this behavior!

Reply:
Of course, we could also just set it to 4 here in the Analyzers.targets if the project is a .NETFramework project, which I've done in the interim here as an example.